libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -2,375 +2,957 @@
+
+    // Estimate rate of descent
+    static const float ASSUMED_CD0 = 0.011;
+    const float sink_rate = ((0.25 * get_solidity() * ASSUMED_CD0 / c_t_hover) + (sq(c_t_hover) / (get_solidity() * ASSUMED_CD0))) * tip_speed_auto;
+
+    // Calc flare altitude
+    // TODO: We need to come up with some way to modify the forward speed target here as the target speed is
+    // body frame forward speed and the flare height calculation needs earth frame forward speed.  The trouble is
+    // that this is a projection into the future so we cannot provide a measured value at this stage
+    float des_spd_fwd = _param_target_speed.get();
+    calc_flare_hgt(des_spd_fwd, -1.0 * sink_rate);
+
+    gcs().send_text(MAV_SEVERITY_INFO, "Ct/sigma=%f W=%f kg flare_alt=%f", c_t_hover/get_solidity(), _hover_thrust/GRAVITY_MSS, _flare_hgt.get());
+    gcs().send_text(MAV_SEVERITY_INFO, "sink rate=%f", sink_rate);
+    gcs().send_text(MAV_SEVERITY_INFO, "inflow spd=%f", lambda_eq*tip_speed_auto);
+}
+
+void AC_Autorotation::calc_flare_hgt(const float fwd_speed, float climb_rate)
+{
+    // we must be descending for this maths to sensible
+    if (!is_negative(climb_rate)) {
+        return;
+    }
+
+    // Keep the slow filter of vel z up to date. We always update this with a fresh measurement as the
+    // climb_rate argument maybe from an approximated climb rate not a measured one
+    _lagged_vel_z.apply(get_ef_velocity_up(), _dt);
 
-    if (_flags.bad_rpm) {
-        //count unhealthy rpm updates and reset healthy rpm counter
-        _unhealthy_rpm_counter++;
-        _healthy_rpm_counter = 0;
+    // Estimate total rotor drag force coefficient in the descent
+    // This is not the fully non-dimensionalized drag force to avoid having to constantly
+    // dividing and multiply vehicle constants like rotor diameter
+    float CR = _hover_thrust / sq(climb_rate);
+    CR = constrain_float(CR, 0.01, 1.7); // TODO: confirm typical range of drag coefficients expected
 
-    } else if (!_flags.bad_rpm && _unhealthy_rpm_counter > 0) {
-        //poor rpm reading may have cleared.  Wait 10 update cycles to clear.
-        _healthy_rpm_counter++;
+    // Compute speed module and glide path angle during descent
+    const float speed_module = MAX(norm(climb_rate, fwd_speed), 0.1); // (m/s)
+    const float glide_angle = M_PI / 2 - safe_asin(fwd_speed / speed_module); // (rad)
 
-        if (_healthy_rpm_counter >= 10) {
-            //poor rpm health has cleared, reset counters
-            _unhealthy_rpm_counter = 0;
-            _healthy_rpm_counter = 0;
-        }
+    // Estimate inflow velocity at beginning of flare
+    float entry_inflow = - speed_module * sinf(glide_angle + radians(AP_ALPHA_TPP));
+
+    const float k_1 = safe_sqrt(_hover_thrust / CR); // TODO: discuss: in the initial estimate k1 will always equal climb rate, see definition of CR above
+
+    // Protect against div by 0 case
+    if (is_zero(climb_rate + k_1)) {
+        climb_rate -= 0.05;
```

Comment:
something I noticed in the rework/simplification.  climb_rate always equals k_1 because of the way k_1 is calculated from rotor drag (CR) and hover thrust.  See lines 660 and 650.  It was less obvious before because the calc for CR was always called externally to this function.  
This means that the denominator in the calc for "a" (line 674) is always this arbitrary -0.05 value.